### PR TITLE
Dynamic coin change implemented

### PR DESCRIPTION
**Pull request by kolosovpetro**
- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines of this project
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Comments in areas I changed are up to date
- [x] I have added comments to hard-to-understand areas of my code
- [x] I have made corresponding changes to the README.md

### CODE REVIEW

File: Algorithms/Algorithms.csproj [hunk 1]
```diff
@@ -28,4 +28,8 @@
     <ProjectReference Include="..\Utilities\Utilities.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Folder Include="Problems\CoinChange" />
+  </ItemGroup>
+
```
Comment (siriak) [inline Algorithms/Algorithms.csproj:34]: This is redundant

File: Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    public static class DynamicCoinChangeSolver
+    {
+        public static int[] GenerateSingleCoinChanges(int coin, IEnumerable<int> coins)
+        {
+            ValidateCoin(coin);
+
+            // ReSharper disable once PossibleMultipleEnumeration
+            ValidateCoinsArray(coins);
+
+            // ReSharper disable once PossibleMultipleEnumeration
```
Comment (siriak) [inline Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs:16]:
```suggestion
        public static int[] GenerateSingleCoinChanges(int coin, int[] coins)
        {
            ValidateCoin(coin);

            ValidateCoinsArray(coins);

```

File: Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+        private static void ValidateCoin(int coin)
+        {
+            if (coin <= 0)
+            {
+                throw new InvalidOperationException($"Coin cannot be lesser or equal to zero {nameof(coin)}.");
+            }
+        }
+
+        private static void ValidateCoinsArray(IEnumerable<int> enumerableCoins)
```
Comment (siriak) [inline Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs:114]: It says array, but accepts IEnumerable

File: Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            {
+                throw new InvalidOperationException($"Coins array must contain coin 1 {nameof(coinsAsArray)}.");
+            }
+
+            var containsNegative = coinsAsArray.Any(x => x <= 0);
+
+            if (containsNegative)
+            {
+                throw new InvalidOperationException(
+                    $"Coins array cannot contain negative numbers {nameof(coinsAsArray)}.");
```
Comment (siriak) [inline Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs:140]: What about 0?

File: Algorithms.Tests/Problems/CoinChange/Dynamic/GenerateSingleCoinChangesTests.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        public void GenerateSingleCoinChangesTests_ShouldThrow_CoinsArrayCannotBeNull()
+        {
+            const int coin = 10;
+            int[] coinsAsArray = default!;
+
+            Func<int[]> act = () => DynamicCoinChangeSolver.GenerateSingleCoinChanges(coin, coinsAsArray);
+
+            act.Should().Throw<ArgumentNullException>()
+                .WithMessage("Value cannot be null. (Parameter 'enumerableCoins')");
+        }
```
Comment (siriak) [inline Algorithms.Tests/Problems/CoinChange/Dynamic/GenerateSingleCoinChangesTests.cs:58]: No need in this test and the null check, the method signature clearly says that nulls are not allowed

File: Algorithms.Tests/Problems/CoinChange/Dynamic/GenerateSingleCoinChangesTests.cs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        [Test]
+        public void GenerateSingleCoinChangesTests_ShouldThrow_CoinCannotBeLesserOrEqualZero()
+        {
+            const int coin = 0;
+            var arr = new[] { 1, 2, 3 };
+
+            Func<int[]> act = () => DynamicCoinChangeSolver.GenerateSingleCoinChanges(coin, arr);
+
+            act.Should().Throw<InvalidOperationException>()
+                .WithMessage($"Coin cannot be lesser or equal to zero {nameof(coin)}.");
```
Comment (siriak) [inline Algorithms.Tests/Problems/CoinChange/Dynamic/GenerateSingleCoinChangesTests.cs:45]:
```suggestion
                .WithMessage($"The coin cannot be less than or equal to zero {nameof(coin)}.");
```

**Review by siriak — CHANGES_REQUESTED** [review]
Overall looks good, please add some comments explaining what methods do

File: Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs (hunk 4 above)
Comment (kolosovpetro) [reply Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs:140]: Current implementation have a few requirements. 

First, it should always contain an item 1, just because of divisibility, for instance we cannot exchange 12 having coins = [3,5], because remainder 2 is lesser than minimum 3. Keeping an item 1 in the array we are able to exchange any integer.

About to keep 0, it does not make sense since that any number except 0 cannot be exchanged by 0, so it that case, for current implementation, it will be an infinit loop.

File: Algorithms/Algorithms.csproj [hunk 7]
```diff
@@ -27,5 +27,5 @@
     <ProjectReference Include="..\DataStructures\DataStructures.csproj" />
     <ProjectReference Include="..\Utilities\Utilities.csproj" />
   </ItemGroup>
-
+
```
Comment (siriak) [inline Algorithms/Algorithms.csproj:30]: Please remove the spaces

File: Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+                throw new InvalidOperationException($"The coin cannot be lesser or equal to zero {nameof(coin)}.");
+            }
+        }
+
+        private static void ValidateCoinsArray(int[] coinsArray)
+        {
+            if (coinsArray == null)
+            {
+                throw new ArgumentNullException(nameof(coinsArray));
+            }
```
Comment (siriak) [inline Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs:147]: Please remove this check, int[] is not nullable. I get that the user can technically still pass null there, but it's their problem. The interface states clearly that coins are not nullable.

**Review by siriak — COMMENTED** [review]
Looks good, only minor suggestions

File: Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+
+            var coinsContainOne = coinsAsArray.Any(x => x == 1);
+
+            if (!coinsContainOne)
+            {
+                throw new InvalidOperationException($"Coins array must contain coin 1 {nameof(coinsAsArray)}.");
+            }
+
+            var containsNegative = coinsAsArray.Any(x => x <= 0);
```
Comment (siriak) [inline Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs:158]:
```suggestion
            var containsNonPositive = coinsAsArray.Any(x => x <= 0);
```

File: Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+
+            if (containsNegative)
```
Comment (siriak) [inline Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs:160]:
```suggestion
            if (containsNonPositive)
```

File: Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+            {
+                throw new InvalidOperationException(
+                    $"Coins array cannot contain the numbers lesser than zero {nameof(coinsAsArray)}.");
```
Comment (siriak) [inline Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs:163]:
```suggestion
                    $"{nameof(coinsAsArray)} cannot contain the numbers less than or equal to zero");
```

File: Algorithms.Tests/Problems/CoinChange/Dynamic/GenerateSingleCoinChangesTests.cs [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        [Test]
+        public void GenerateSingleCoinChangesTests_ShouldThrow_CoinsArrayCannotContainNegativeValues()
+        {
+            const int coin = 10;
+            var coinsAsArray = new[] { 1, 2, -3, 4 };
+
+            Func<int[]> act = () => DynamicCoinChangeSolver.GenerateSingleCoinChanges(coin, coinsAsArray);
+
+            act.Should().Throw<InvalidOperationException>()
+                .WithMessage($"Coins array cannot contain the numbers lesser than zero {nameof(coinsAsArray)}.");
```
Comment (siriak) [inline Algorithms.Tests/Problems/CoinChange/Dynamic/GenerateSingleCoinChangesTests.cs:81]:
```suggestion
                .WithMessage($"{nameof(coinsAsArray)} cannot contain numbers less than or equal to zero");
```

File: Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+            {
+                throw new InvalidOperationException($"Coins array must contain coin 1 {nameof(coinsAsArray)}.");
+            }
+
+            var containsNonPositive = coinsAsArray.Any(x => x <= 0);
+
+            if (containsNonPositive)
+            {
+                throw new InvalidOperationException(
+                    $"{nameof(coinsAsArray)} cannot contain the numbers less than or equal to zero");
```
Comment (siriak) [inline Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs:163]:
```suggestion
                    $"{nameof(coinsAsArray)} cannot contain numbers less than or equal to zero");
```

**Review by siriak — APPROVED** [review]
LGTM, thanks!
